make pillar top_matches work with node groups

mattp- opened the pull request:
### What does this PR do?
the compound match evaluator to date relied callers to pre-evaluate nodegroups
into their decomposed state. in some places this wasn't happening. Rather than
fix every call site, teach compound_match to evaluate nodegroups in place.

It seems matchers were refactored to salt.utils.minions at some point but the
original minion.py matcher was not ported to use it. Given the hugeness of that
change, i've left as is and ported my fix to salt.utils.minions as well
instead.

### What issues does this PR fix or reference?
this should fix #6602

### Previous Behavior
nodegroup in compound didn't work.

### New Behavior
it works.

### Tests written?
Yes

### Commits signed with GPG?
No

Please review [Salt's Contributing Guide](https://docs.saltstack.com/en/latest/topics/development/contributing.html) for best practices.

See GitHub's [page on GPG signing](https://help.github.com/articles/signing-commits-using-gpg/) for more information about signing commits with GPG.

mattp- committed:
make pillar top_matches work with node groups

the compound match evaluator to date relied callers to pre-evaluate nodegroups
into their decomposed state. in some places this wasn't happening. Rather than
fix every call site, teach compound_match to evaluate nodegroups in place.

It seems matchers were refactored to salt.utils.minions at some point but the
original minion.py matcher was not ported to use it. Given the hugeness of that
change, i've left as is and ported my fix to salt.utils.minions as well
instead.

isbm commented on salt/minion.py:
```diff
@@ -3415,10 +3419,11 @@ def compound_match(self, tgt):
-                    return False
+                    # if we encounter a node group, just evaluate it in-place
+                    decomposed = salt.utils.minions.nodegroup_comp(target_info['pattern'], nodegroups)
+                    words = decomposed + words
```
You can do it with less coding:
```python
words[0:0] = salt.utils.minions.nodegroup_comp(target_info['pattern'], nodegroups)
```

mattp- committed:
account for failed expansions of nodegroups

we may receive '' empty-string for nodegroup expansions, so we must account for
that safely.

mattp- committed:
add some tests for compound nodegroup matching

mattp- committed:
update docs to reference new feature

mattp- commented:
@rallytime i believe this should be good to go now

Commit:
Lint: Remove extra blank line

corywright commented:
Dreams **can** come true!

Thanks @mattp-!
